hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECKeyOutputStream.java

```diff
@@ -259,7 +259,7 @@ private StripeWriteStatus rewriteStripeToNewBlockGroup(int chunkSize,
         blockOutputStreamEntryPool.getCurrentStreamEntry();
     newBlockGroupStreamEntry
         .updateBlockGroupToAckedPosition(failedStripeDataSize);
-    ecChunkBufferCache.clear(chunkSize);
```

Comment:
writeParityCells allocating the buffers based on the current max chunk size in that stripe. Lets say, the stripe smaller, then the chunksize for parity could be smaller then default chunk size. So, we are allocating buffer with that size. I think this will polute the bufferpool by creating differed size buffers in it. Since we are using the pool globally now, we may need to create same sized buffers and limit the sizes as needed?

Reply:
There seems to be two places in the code where we "allocateParityBuffers". When finishing a stripe and when finishing the file. When finishing the stripe, we always ask for a ecChunkSize buffers, but when calling close, we ask for buffers for the desired size. I think we should always ask for the full chunksize, and then set the limits accordingly.

However looking at the `allocateBuffers()` code:

```
    private void allocateBuffers(ByteBuffer[] buffers, int bufferSize) {
      for (int i = 0; i < buffers.length; i++) {
        buffers[i] = byteBufferPool.getBuffer(false, cellSize);
        buffers[i].limit(bufferSize);
      }
    }
```

We always allocate the cellSize, but then set the limit, so I think we always ask for ecChunkSize buffers.

Looking into this code a bit more, I see we always call `allocateParityBuffers()` in `writeParityCells()`. Is this the correct thing to do? Should we not re-use the parity buffers for the life of the block, and just clear them between stripes?

Also, do we release the parity buffers, before allocating them again, if we allocate them on each stripe write?

From the above code snippet, if we call `allocateParityBuffers()`:

```
    public void allocateParityBuffers(int size){
      allocateBuffers(parityBuffers, size);
    }
```

It will just overwrite the slots in the parityBuffer array, which will leak buffers in the pool.

Am I reading the code correctly, or am I missing something?

Prior to this PR, the pool is a static variable - if I am correct about the above, then I think the leak exists even without this change?

Reply:
Yeah you are right. There is bug exist. I think we started to created both buffers at once and just get the buffers whenever we need. We should make allocateBuffers API private and use get buffers. AllocateBuffers can happen in ctor.
Originally we started with that: https://github.com/umamaheswararao/hadoop-ozone/blob/e82f53ee89c8e2fbe3c4fd4157958180f3621ead/hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECKeyOutputStream.java

Eventually started to allocateBuffers, probably to make that diffred cellSize buffers to work.

So, the simpler fix would be to replace allocateBuffers with getBuffers and make the buffers allocated  in ctor.
Once we get the buffers we can set the right limits. If you want to cover here please go ahead Otherwise you can file one JIRA and assign it to me.

Reply:
I believe something like this should fix that existing issue of buffer overriding on allocateBuffers in writeParityCells:
```
--- a/hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECKeyOutputStream.java
+++ b/hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECKeyOutputStream.java
@@ -319,8 +319,10 @@ private void addToExcludeNodesList(List<ECBlockOutputStream> failedStreams) {
 
   void writeParityCells(int parityCellSize) throws IOException {
     final ByteBuffer[] buffers = ecChunkBufferCache.getDataBuffers();
-    ecChunkBufferCache.allocateParityBuffers(parityCellSize);
     final ByteBuffer[] parityBuffers = ecChunkBufferCache.getParityBuffers();
+    for (int i=0;i<parityBuffers.length;i++){
+      parityBuffers[i].limit(parityCellSize);
+    }
 
     for(int i=0; i< buffers.length; i++){
       buffers[i].flip();
@@ -639,6 +641,7 @@ private void checkNotClosed() throws IOException {
       dataBuffers = new ByteBuffer[numData];
       parityBuffers = new ByteBuffer[numParity];
       allocateBuffers(cellSize, dataBuffers);
+      allocateBuffers(cellSize, parityBuffers);
     }
 
     private ByteBuffer[] getDataBuffers() {
@@ -649,10 +652,6 @@ private void checkNotClosed() throws IOException {
       return parityBuffers;
     }
 
-    public void allocateParityBuffers(int size){
-      allocateBuffers(size, parityBuffers);
-    }
-
```

Reply:
I have pushed another commit based on the diff above to hopefully resolve the issue around buffers leaking.